pennylane_lightning/core/src/simulators/lightning_kokkos/measurements/MeasuresFunctors.hpp

```diff
@@ -16,27 +16,455 @@
+template <class PrecisionT, class DeviceType> class getProbsFunctor {
+  public:
+    // Required for functor:
+    using execution_space = DeviceType;
+    using value_type = PrecisionT[];
+    const unsigned value_count;
 
-    getProbFunctor(Kokkos::View<Kokkos::complex<PrecisionT> *> arr_,
-                   Kokkos::View<PrecisionT *> probability_)
-        : arr(arr_), probability(probability_) {}
+    using ComplexT = Kokkos::complex<PrecisionT>;
+    Kokkos::View<ComplexT *> arr;
+    Kokkos::View<std::size_t *> all_indices;
+    Kokkos::View<std::size_t *> all_offsets;
+    getProbsFunctor(const Kokkos::View<ComplexT *> &arr_,
+                    const std::vector<std::size_t> &wires_,
+                    const Kokkos::View<std::size_t *> all_indices_,
+                    const Kokkos::View<std::size_t *> all_offsets_)
+        : value_count{1U << wires_.size()}, arr{arr_},
+          all_indices{all_indices_}, all_offsets{all_offsets_} {}
 
     KOKKOS_INLINE_FUNCTION
-    void operator()(const std::size_t k) const {
-        const PrecisionT REAL = arr(k).real();
-        const PrecisionT IMAG = arr(k).imag();
-        probability(k) = REAL * REAL + IMAG * IMAG;
+    void init(PrecisionT dst[]) const {
+        for (unsigned i = 0; i < value_count; ++i)
+            dst[i] = 0.0;
+    }
+
+    KOKKOS_INLINE_FUNCTION
+    void join(PrecisionT dst[], const PrecisionT src[]) const {
+        for (unsigned i = 0; i < value_count; ++i)
+            dst[i] += src[i];
+    }
+
+    KOKKOS_INLINE_FUNCTION
+    void operator()(const std::size_t i, const std::size_t j,
+                    PrecisionT dst[]) const {
+        const std::size_t index = all_indices(i) + all_offsets(j);
+        const PrecisionT rsv = arr(index).real();
+        const PrecisionT isv = arr(index).imag();
+        dst[i] += rsv * rsv + isv * isv;
+    }
+};
+
+/**
+ * @brief Compute probability distribution from StateVector.
+ *
+ * @tparam PrecisionT StateVector precision.
+ * @tparam DeviceType Kokkos execution space.
+ * @tparam num_wires Number of wires (0 is used for a dynamic number of wires).
+ * @param arr_ StateVector data.
+ * @param num_qubits_ Number of qubits.
+ * @param wires_ Wires for which the probability is computed.
+ */
+template <class PrecisionT, class DeviceType, std::size_t num_wires>
+class getProbsNQubitOpFunctor {
+  public:
+    // Required for functor:
+    using execution_space = DeviceType;
+    using value_type = PrecisionT[];
+    const unsigned value_count;
+
+    using UnmanagedSizeTHostView =
+        Kokkos::View<std::size_t *, Kokkos::HostSpace,
+                     Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+    using ComplexT = Kokkos::complex<PrecisionT>;
+    using KokkosComplexVector = Kokkos::View<ComplexT *>;
+    Kokkos::View<ComplexT *> arr;
+    const std::size_t n_wires;
+
+    Kokkos::View<std::size_t *> parity;
+    Kokkos::View<std::size_t *> rev_wires;
+
+    std::size_t rev_wire_0;
```

Comment:
Can we put variables here into a `std::array` object or an array[]?

Reply:
No, because `std::array` is a host object and hence it won't work on GPU devices. This is also necessary for performance.

Reply:
What about `rev_wire[9]`?

Reply:
We don't use `rev_wire[9]` explicitly anywhere. For more than 8 wires, we switch to using Kokkos views and loops which generic but slower.